compiler/rustc_mir/src/transform/check_consts/qualifs.rs

```diff
@@ -34,15 +37,21 @@ pub fn in_any_value_of_ty(
 /// To accomplish this, const-checking and promotion use a value-based analysis (as opposed to a
 /// type-based one). Qualifications propagate structurally across variables: If a local (or a
 /// projection of a local) is assigned a qualifed value, that local itself becomes qualifed.
-pub trait Qualif {
+pub(crate) trait Qualif {
     /// The name of the file used to debug the dataflow analysis that computes this qualif.
     const ANALYSIS_NAME: &'static str;
 
+    /// The dataflow result type. If it's just qualified/not qualified, then
+    /// you can just use a `()` (most qualifs do that). But if you need more state, use a
+    /// custom enum.
+    type Result: SetChoice + std::fmt::Debug = ();
```

Comment:
I think I would prefer to use a marker ZST that is individual to each qualif (or even just use the type implementing `Qualif`), but since that is a refactoring not necessary for this PR, I think we should not do it here.